Source/CarthageKit/Project.swift

```diff
@@ -1289,14 +1290,78 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
 			}
 			return frameworksInURL.count == 1
 		}.filter { url in
-
-			let packageType: PackageType? = Bundle(url: url)?.packageType
-
+			
+			// Try reading the Info.plist of the Framework
+			let bundle = Bundle(url: url)
+			let packageType: PackageType? = bundle?.packageType
+			
 			switch packageType {
 			case .framework?, .bundle?:
 				return true
 			default:
-				return false
+				// In case no Info.plist exists check the Mach-O fileType
+				guard let executablePath = bundle?
+					.executableURL?
+					.resolvingSymlinksInPath()
+					.path else {
+						
+					return false
+				}
+				
+				// Read the Mach header with `objdump` and parse the output.
+				// The output is composed of one or more sets of lines like the following:
+				//
+				// Mach header
+				//      magic cputype  cpusubtype  caps    filetype ncmds sizeofcmds      flags
+				// 0xfeedfacf 16777223          3  0x00           1     8       1720 0x00002000
+				//
+				// See output structure at:
+				// https://llvm.org/viewvc/llvm-project/llvm/trunk/tools/llvm-objdump/MachODump.cpp?view=markup&pathrev=225383###see%C2%B7line%C2%B72745
+
+				let task = Task("/usr/bin/xcrun", arguments: [
+					"objdump",
+					"-macho",
+					"-private-header",
+					"-non-verbose",
+					executablePath
+					]
+				)
+				let fileType = task.launch(standardInput: nil)
+					.ignoreTaskData()
+					.map { String(data: $0, encoding: .utf8) ?? "" }
+					.filter { !$0.isEmpty }
+					.flatMap(.merge) { (output: String) -> SignalProducer<String, NoError> in
+						output.linesProducer
+					}.filter {
+						// Keep the lines with Mach-O magic number indefiers
+						$0.contains(String(format: "%x", MH_CIGAM))
+							|| $0.contains(String(format: "%x", MH_MAGIC))
+							|| $0.contains(String(format: "%x", MH_CIGAM_64))
+							|| $0.contains(String(format: "%x", MH_MAGIC_64))
+					}
+					.take(last: 1)
+					.filterMap { (magicLine: String) -> Int? in
+						
+						let components = magicLine
+							.components(separatedBy: CharacterSet.whitespaces)
+							.filter { !$0.isEmpty }
+						// A Mach-O header has 8 components, see example lines in comments above
+						guard components.count == 8 else {
+							return nil
+						}
+						
+						// The 4th component is the fileType
+						return Int(components[4])
+					}
+					.filter {
+						$0 == MH_OBJECT // Carthage accepts static libraries
+						|| $0 == MH_BUNDLE // Bundles https://github.com/ResearchKit/ResearchKit/blob/1.3.0/ResearchKit/Info.plist#L15-L16
+						|| $0 == MH_DYLIB // or dynamic shared libraries
+					}
+					.single()?
+					.value
+
+					return fileType != nil
```

Comment:
I think this code ought to be split out into its own function.

Reply:
I'd want to see tests for this as well.

Reply:
ok